src/backend/distributed/executor/local_executor.c

```diff
@@ -0,0 +1,579 @@
+ *		nodes.
+ *
+ *		Note that for read-only queries, after the local execution, there is no need to
+ *		kick in adaptive executor.
+ *
+ *  There are also few limitations/trade-offs that is worth mentioning. First, the
+ *  local execution on multiple shards might be slow because the execution has to
+ *  happen one task at a time (e.g., no parallelism). Second, if a transaction
+ *  block/CTE starts with a multi-shard command, we do not use local query execution
+ *  since local execution is sequential. Basically, we do not want to lose parallelism
+ *  across local tasks by switching to local execution. Third, the local execution
+ *  currently only supports queries. In other words, any utility commands like TRUNCATE,
+ *  fails if the command is executed after a local execution inside a transaction block.
+ *  Forth, the local execution cannot be mixed with the executors other than adaptive,
+ *  namely task-tracker, real-time and router executors. Finally, related with the
+ *  previous item, COPY command cannot be mixed with local execution in a transaction.
+ *  The implication of that any part of INSERT..SELECT via coordinator cannot happen
+ *  via the local execution.
+ */
+#include "postgres.h"
+#include "miscadmin.h"
+
+#include "distributed/citus_custom_scan.h"
+#include "distributed/local_executor.h"
+#include "distributed/multi_executor.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/relation_access_tracking.h"
+#include "distributed/remote_commands.h" /* to access LogRemoteCommands */
+#include "distributed/transaction_management.h"
+#include "executor/tstoreReceiver.h"
+#include "executor/tuptable.h"
+#include "optimizer/planner.h"
+#include "nodes/params.h"
+#include "utils/snapmgr.h"
+
+
+/* controlled via a GUC */
+bool EnableLocalExecution = true;
+bool LogLocalCommands = false;
+
+bool LocalExecutionHappened = false;
+
+
+static void SplitLocalAndRemoteTasks(bool readOnlyPlan, List *taskList,
+									 List **localTaskList, List **remoteTaskList);
+static void SplitLocalAndRemotePlacements(List *taskPlacementList,
+										  List **localTaskPlacementList,
+										  List **remoteTaskPlacementList);
+static uint64 ExecuteLocalTaskList(CitusScanState *node, List *taskList);
+static Oid * GetParamListOids(ParamListInfo paramList);
+static uint64 ExecuteLocalTaskPlan(CitusScanState *scanState, PlannedStmt *taskPlan,
+								   char *queryString);
+static void LogLocalCommand(const char *command);
+
+
+/*
+ * ExecuteLocalTasks gets a CitusScanState node and executes any tasks that are
+ * local to the node executing the query.
+ *
+ * The function returns true if at least one task has been locally executed. The
+ * function also fills the remoteTaskList so that the caller (e.g., adaptive executor)
+ * can continue with executing the remote tasks.
+ */
+bool
+ExecuteLocalTasks(CitusScanState *node, List **remoteTaskList)
```

Comment:
I find the naming here a bit confusing. I kind of expect we execute some set of local tasks (like ExecuteLocalTaskList), but it's rather that we'll first determine what work can be done locally and construct a task list from that and then also execute it.

Anyway, would be nice to move the building of separate task lists into StartDistributedExecution.